### PR TITLE
Added distinct to NonEmptyList

### PR DESCRIPTION
**Pull request by Michaelt293**
Feel free to suggest a more performant implementation :)

### CODE REVIEW

**Comment by adamgfraser** [comment]
@Michaelt293 Thanks!

The problem with this is that it is going to have N squared complexity because for each element in the collection we have to traverse the list of accumulated values. To avoid this we need a way to determine whether an element is in the set of value we have seen before that doesn't require potentially comparing every element, which requires a way of hashing the values. How about something like this?

```scala
final def distinct(implicit A: Hash[A]): NonEmptyList[A] =
  reduceMapLeft(a => (single(a), Set(a.hash))) { case ((as, seen), a) =>
    val hash = a.hash
    if (seen(hash)) (as, seen) else (cons(a, as), seen + hash)
  }._1.reverse
```

**Comment by Michaelt293** [comment]
Thanks @adamgfraser, I was considering passing around a set to keep track of the seen elements. I will go ahead and make the change.

File: src/main/scala/zio/prelude/NonEmptyList.scala [hunk 1]
```diff
@@ -54,6 +54,16 @@ sealed trait NonEmptyList[+A] { self =>
...
     foldLeft(0)((n, a) => if (f(a)) n + 1 else n)
 
+  /**
+   * Removes duplicate elements from this `NonEmptyList`.
+   */
+  final def distinct(implicit A: Hash[A]): NonEmptyList[A] =
+    reduceMapLeft(a => (single(a), Set(a.hash))) {
+      case ((as, seen), a) =>
+        val hash = a.hash
+        if (seen(hash)) (as, seen) else (cons(a, as), seen + hash)
```
Comment (joroKr21) [inline src/main/scala/zio/prelude/NonEmptyList.scala:64]: Hmm what if there is a hash collision? 🤔

Comment (Michaelt293) [reply src/main/scala/zio/prelude/NonEmptyList.scala:64]: If hash collisions are a concern, a `Equal` constraint could be used instead. @adamgfraser let me know if you want me to raise a PR.

**Pull request merged**